### DISCUSSION TITLE
Ability to import address spaces non-Azure

### DISCUSSION THREAD

**Discussion by cata008** [Azure/ipam#118] [General]
Hi guys,

First of all congrats for the IPAM project, can't wait for it to become a native solution available in Marketplace.

My question is whether you have an ETA at this moment in regards to allowing the IPAM tool to import non-Azure ranges? 

My client uses a VWAN infra with quite a few site-to-sites and one ExpressRoute. For such use cases, it is quite imperative to have visibility over the IP address ranges that have already been used by each branch.

Cheers!

**Comment by tonydean37** [comment]
Hi,

I just saw this discussion after coming across a challenge with our setup regarding having vnets connected to our vwan from another tenant. So posting here rather than setting up a new chat (@DCMattyG - let me know if you would prefer a new post :) ). Unfortunately, it does not get picked up by IPAM, so the same subnet could be assigned to a new vnet. If I could either request a reservation with that specific subnet, or (I think preferably) add a custom entry for it, then that would be great. What do you think?

**Comment by DCMattyG** [comment]
Hi @cata008 and @tonydean37, I'm actually starting to work on this now!

The idea would be "static CIDR's" that you could assign to Blocks, representing these remote (non-Azure) networks. Would something like that meet your needs?

**Reply by cata008** [reply]
Thanks for your quick response @DCMattyG. I believe that would work for us.

**Reply by DCMattyG** [reply]
Excellent @cata008. I'll get to work, and share my progress here for you and @tonydean37 to review & test!

**Reply by tonydean37** [reply]
Same here - that sounds great thanks :)

**Reply by cata008** [reply]
Hi @DCMattyG. Any ETA on the ability to import non-Azure IP address ranges?

Also, totally unrelated, but I was exploring the idea of deploying the infrastructure using container apps. I know this is on your roadmap but since the ui and engine apps are containerized, I was wondering if there are any dependencies I might overlook and thus stick better to using app services instead. Thanks in advance!

**Reply by cata008** [reply]
@DCMattyG - any update on this?

**Reply by DCMattyG** [reply]
Hi there @cata008, I'm so sorry for the delay getting back to you on this. I'm hoping to having this feature complete by the end of the month, and I will certainly reach out to you for beta testing & verification that the new capability meets your needs prior to pushing it into production.

**Reply by cata008** [reply]
Thanks for the update @DCMattyG ! :)

**Reply by cata008** [reply]
Hi @DCMattyG, do you have an update on this or a new ETA? Cheers!

**Reply by cata008** [reply]
Hi @DCMattyG. We are about to implement the IPAM solution. Did you get to work on this, any update? Thanks so much!

**Reply by DCMattyG** [reply]
Hey @cata008, my apologies for being so unresponsive. It's the end of the FY here at Microsoft and everything is crazy!

Quick update...as I was beginning work on adding this feature, I identified a few areas of "technical debt" that needed to be re-worked to better accommodate this functionality. I just pushed a PR with tons of updates, and my currently priority is this item.

I will add this ability via API-only at first, just to get it out the door ASAP. Shortly thereafter, I'll have this feature visible within the UI as well. Hopefully the API-only one will be out by the end of June (~1 week), and the full version within the UI will be ~2 weeks after that (ETA July 17th).

Again, sorry for the delays, but we're almost there. You'll be the first to know when this becomes available for testing!

**Reply by cata008** [reply]
Thanks so much for the update @DCMattyG. Looking forward to hearing back from you! :)

**Reply by DCMattyG** [reply]
Hi @cata008, my apologies for taking a bit longer than expected, but I wanted to update you that I'm nearing completion of this feature. He's a sneak peak at the UI experience when adding a new External Network (CIDR):

<img width="1542" alt="IPAM_External_Routes" src="https://github.com/Azure/ipam/assets/12955732/54933c83-27f6-4419-bd81-79a9f456fecd">

I should have this ready for you to test & verify by the end of this month. This one was a bit complex, but glad to have it near completion and ready for you shortly!

**Reply by cata008** [reply]
Thanks for the update @DCMattyG. Looking forward to testing this one!

**Comment by DCMattyG** [comment]
At last @cata008, the moment you've been waiting for has arrived!

I've completed the work to include External Networks within the Azure IPAM tool. This functionality is currently available on the DEV container images, so here is how you can test everything out...

Go to your Azure App Service which is running your Azure IPAM services, navigate to **Deployment Center**, and update the 3 references to the Azure Container Registry (ACR) to have the suffix "dev" (so they will be "azureipamdev"). Save the changes, and the App Service will restart and download the new containers.

![image](https://github.com/Azure/ipam/assets/12955732/be29c2dd-c1eb-4ce5-92b0-ceb6e4e0ea05)

Once you have the DEV code up and running, here is how you can start to add External Networks...

![image](https://github.com/Azure/ipam/assets/12955732/77d51cea-cf47-43b0-a46a-4fbd8b2fd090)

![image](https://github.com/Azure/ipam/assets/12955732/71b64479-2cf7-470c-8c6e-c306aa3012a3)

![image](https://github.com/Azure/ipam/assets/12955732/c5cefec1-466e-42c5-ac6e-4ebe2d2ab381)

There are also a full suite of API's for this functionality as well:

![image](https://github.com/Azure/ipam/assets/12955732/d7505d56-9443-44ee-a610-5389505995a7)

Here are the rules:
1. You cannot add an External Network to a Block that is outside of the Block's CIDR
2. You cannot add an External Network that overlaps a Virtual Network currently associated to the Block
3. You cannot add an External Network that overlaps an active (non-settled) Reservation on the Block

So, I really hope you enjoy! Please give me feedback (good & bad), and once we're sure this is working as expected, I'll push the code into production.

Thanks so much for your patience and support!

**Reply by DCMattyG** [reply]
Hi @cata008, just wanted to let you know this feature has been merged into Main with #153. Please let me know if you have any feedback (good or bad) and I'm happy to adjust/tweak this as necessary if you have any updates you'd like to see. Have a great one!

**Reply by cata008** [reply]
Hi @DCMattyG. Thanks for the update! I have been off for a while, enjoying my vacation, and unfortunately didn't have time to test the feature. 
I'll be missing again for a couple of weeks but I am eager to pick this up in September. Will keep you posted! :)
